tests/components/zwave/test_init.py

```diff
@@ -1310,699 +1235,851 @@ def mock_connect(receiver, signal, *args, **kwargs):
-        assert len(self.zwave_network.controller.remove_node.mock_calls) == 1
-
-    def test_cancel_command(self):
-        """Test zwave cancel_command service."""
-        self.hass.services.call("zwave", "cancel_command", {})
-        self.hass.block_till_done()
-
-        assert self.zwave_network.controller.cancel_command.called
-        assert len(self.zwave_network.controller.cancel_command.mock_calls) == 1
-
-    def test_heal_network(self):
-        """Test zwave heal_network service."""
-        self.hass.services.call("zwave", "heal_network", {})
-        self.hass.block_till_done()
-
-        assert self.zwave_network.heal.called
-        assert len(self.zwave_network.heal.mock_calls) == 1
-
-    def test_soft_reset(self):
-        """Test zwave soft_reset service."""
-        self.hass.services.call("zwave", "soft_reset", {})
-        self.hass.block_till_done()
-
-        assert self.zwave_network.controller.soft_reset.called
-        assert len(self.zwave_network.controller.soft_reset.mock_calls) == 1
-
-    def test_test_network(self):
-        """Test zwave test_network service."""
-        self.hass.services.call("zwave", "test_network", {})
-        self.hass.block_till_done()
-
-        assert self.zwave_network.test.called
-        assert len(self.zwave_network.test.mock_calls) == 1
-
-    def test_stop_network(self):
-        """Test zwave stop_network service."""
-        with patch.object(self.hass.bus, "fire") as mock_fire:
-            self.hass.services.call("zwave", "stop_network", {})
-            self.hass.block_till_done()
-
-            assert self.zwave_network.stop.called
-            assert len(self.zwave_network.stop.mock_calls) == 1
-            assert mock_fire.called
-            assert len(mock_fire.mock_calls) == 1
-            assert mock_fire.mock_calls[0][1][0] == const.EVENT_NETWORK_STOP
-
-    def test_rename_node(self):
-        """Test zwave rename_node service."""
-        self.zwave_network.nodes = {11: MagicMock()}
-        self.hass.services.call(
-            "zwave",
-            "rename_node",
-            {const.ATTR_NODE_ID: 11, const.ATTR_NAME: "test_name"},
-        )
-        self.hass.block_till_done()
+async def test_device_config_glob_is_ordered():
+    """Test that device_config_glob preserves order."""
+    conf = CONFIG_SCHEMA({"zwave": {CONF_DEVICE_CONFIG_GLOB: OrderedDict()}})
+    assert isinstance(conf["zwave"][CONF_DEVICE_CONFIG_GLOB], OrderedDict)
 
-        assert self.zwave_network.nodes[11].name == "test_name"
 
-    def test_rename_value(self):
-        """Test zwave rename_value service."""
-        node = MockNode(node_id=14)
-        value = MockValue(index=12, value_id=123456, label="Old Label")
-        node.values = {123456: value}
-        self.zwave_network.nodes = {11: node}
+async def test_add_node(hass, mock_openzwave, zwave_setup):
+    """Test zwave add_node service."""
+    zwave_network = hass.data[DATA_NETWORK]
```

Comment:
We should be able to get the network from the `mock_openzwave` mock, ie `mock_openzwave.network.ZWaveNetwork.return_value`.

Reply:
Any idea why it fails?

```
        zwave_network = mock_openzwave.network.ZWaveNetwork.return_value
>       zwave_network.state = MockNetwork.STATE_READY
E       AttributeError: 'property' object has no attribute 'state'
```

Reply:
Not directly. I'd check how the mocking is done.

Reply:
We can improve this later if it's tricky.

Reply:
I'm close to giving up. This is as far as I come.

The `mock_openzwave` patches `openzwave.network`:

```
base_mock.network.ZWaveNetwork = MockNetwork
    base_mock.option.ZWaveOption = MockOption

    with patch.dict(
        "sys.modules",
        {
            "libopenzwave": libopenzwave,
            "openzwave.option": base_mock.option,
            "openzwave.network": base_mock.network,
            "openzwave.group": base_mock.group,
        },
    ):
        yield base_mock
```

In `__init__.py` it creates an instance of ZWaveNetwork:

`network = hass.data[DATA_NETWORK] = ZWaveNetwork(options, autostart=False)`

Is it even possible to get the instance through `mock_openzwave`?

Reply:
I think we'd need to change how we patch the library. We can do that in the future. There's a lot to clean up about the existing zwave tests still.

Reply:
Is there a need to do that? I may clean it up some more, but I need some direction what should be cleaned up.

Generally I think existing tests, at least in `test_init` can be cleaned up further by using the new fixtures introduced. I also found two instances where `caplog` could be used.

Reply:
We should set up the integration and avoid interacting with integration details in the tests.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

The zwave tests are currently not written like that.